api/client/proto/authservice.proto

```diff
@@ -1468,4 +1493,16 @@ service AuthService {
     // the new credentials. When the new authentication is successfully set, any user lock is
     // removed so the user can login immediately afterwards.
     rpc CompleteAccountRecovery(CompleteAccountRecoveryRequest) returns (google.protobuf.Empty);
+
+    // CreateAccountRecoveryCodes (exclusive to cloud users) creates new set of recovery codes for a
+    // user, replacing and invalidating any previously owned codes. Users can only get recovery
+    // codes if their username is in a valid email format.
+    rpc CreateAccountRecoveryCodes(CreateAccountRecoveryCodesRequest)
+        returns (CreateAccountRecoveryCodesResponse);
+    // GetAccountRecoveryToken (exclusive to cloud users) returns a user token resource after
+    // verifying that the token requested has not expired and is of the correct recovery kind.
+    // Besides checking for validity of a token ID, it is also used to get basic information from
+    // the token e.g: username, state of recovery (started or approved) and the type of recovery
+    // requested (password or second factor).
+    rpc GetAccountRecoveryToken(GetAccountRecoveryTokenRequest) returns (types.UserTokenV3);
```

Comment:
Discussed offline, but registering some of the thoughts here.

It looks like someone could try to use this to guess valid recovery tokens. There are some reasons for the RPC to exist (mostly for better UX), but personally I'd rather not have it at all - if the user loses the token for some reason during recovery (closed browser or whatever), then they should start over. One could argue that the circumstances for guessing are rare, and the time window is tight, but I'd rather not have to consider this at all.

As a compromise for keeping it, let's make sure the RPC is rate limited, so at least it is a bit harder to attack.

Reply:
added rate limit